packet/models.py

```diff
@@ -114,65 +116,65 @@ def did_sign(self, username, is_csh):
         # The user must be a misc CSHer that hasn't signed this packet or an off-floor freshmen
         return False
 
-    def is_100(self):
+    def is_100(self) -> bool:
         """
         Checks if this packet has reached 100%
         """
         return self.signatures_required().total == self.signatures_received().total
 
     @classmethod
-    def open_packets(cls):
+    def open_packets(cls) -> list['Packet']:
         """
         Helper method for fetching all currently open packets
         """
         return cls.query.filter(cls.start < datetime.now(), cls.end > datetime.now()).all()
 
     @classmethod
-    def by_id(cls, packet_id):
+    def by_id(cls, packet_id: int) -> 'Packet':
         """
         Helper method for fetching 1 packet by its id
         """
         return cls.query.filter_by(id=packet_id).first()
 
 class UpperSignature(db.Model):
     __tablename__ = 'signature_upper'
-    packet_id = Column(Integer, ForeignKey('packet.id'), primary_key=True)
-    member = Column(String(36), primary_key=True)
-    signed = Column(Boolean, default=False, nullable=False)
-    eboard = Column(String(12), nullable=True)
-    active_rtp = Column(Boolean, default=False, nullable=False)
-    three_da = Column(Boolean, default=False, nullable=False)
-    webmaster = Column(Boolean, default=False, nullable=False)
-    c_m = Column(Boolean, default=False, nullable=False)
-    w_m = Column(Boolean, default=False, nullable=False)
-    drink_admin = Column(Boolean, default=False, nullable=False)
-    updated = Column(DateTime, default=datetime.now, onupdate=datetime.now, nullable=False)
+    packet_id = cast(int, Column(Integer, ForeignKey('packet.id'), primary_key=True))
+    member = cast(str, Column(String(36), primary_key=True))
+    signed = cast(bool, Column(Boolean, default=False, nullable=False))
+    eboard = cast(Optional[str], Column(String(12), nullable=True))
+    active_rtp = cast(bool, Column(Boolean, default=False, nullable=False))
+    three_da = cast(bool, Column(Boolean, default=False, nullable=False))
+    webmaster = cast(bool, Column(Boolean, default=False, nullable=False))
+    c_m = cast(bool, Column(Boolean, default=False, nullable=False))
+    w_m = cast(bool, Column(Boolean, default=False, nullable=False))
+    drink_admin = cast(bool, Column(Boolean, default=False, nullable=False))
+    updated = cast(datetime, Column(DateTime, default=datetime.now, onupdate=datetime.now, nullable=False))
 
-    packet = relationship('Packet', back_populates='upper_signatures')
+    packet = cast(Packet, relationship('Packet', back_populates='upper_signatures'))
```

Comment:
I'm surprised this works

Reply:
Casting is just a no-op at runtime. It's only for the type checker, and it seems to be unconditional.

Reply:
Types on all the imports are a mess, which is why they're ignored completely.

Reply:
There is the `sqlalchemy[mypy]` thing, which definitely made it slightly better, but it doesn't quite work with `flask_sqlalchemy`, so this casting mess is what we're left with.